### PR TITLE
add infrastructure for a -m32 switch

### PR DESCRIPTION
**Pull request by UplinkCoder**
Yep,
I got every oblivious hard-coded 64bit-types and made it possible to vary their bit-widths at runtime.
However there are a few regressions and they are related newExperession. The parameter for `_d_allocmemory` always get's casted to an i64. I spent a few hours tracking down this problem but I can't fix it. EDIT its `LLVMSizeOf`

Regards, Stefan

### CODE REVIEW

File: dub.json [hunk 1]
```diff
@@ -6,8 +6,7 @@
...
...LLVMCppBackendInfo","LLVMMSP430CodeGen","LLVMMSP430Desc","LLVMMSP430Info","LLVMMSP430AsmPrinter","LLVMXCoreDisassembler","LLVMXCoreCodeGen","LLVMXCoreDesc","LLVMXCoreInfo","LLVMXCoreAsmPrinter","LLVMMipsDisassembler","LLVMMipsCodeGen","LLVMMipsAsmParser","LLVMMipsDesc","LLVMMipsInfo","LLVMMipsAsmPrinter","LLVMARMDisassembler","LLVMARMCodeGen","LLVMARMAsmParser","LLVMARMDesc","LLVMARMInfo","LLVMARMAsmPrinter","LLVMAArch64Disassembler","LLVMAArch64CodeGen","LLVMAArch64AsmParser","LLVMAArch64Desc","LLVMAArch64Info","LLVMAArch64AsmPrinter","LLVMAArch64Utils","LLVMPowerPCCodeGen","LLVMPowerPCAsmParser","LLVMPowerPCDesc","LLVMPowerPCInfo","LLVMPowerPCAsmPrinter","LLVMSparcCodeGen","LLVMSparcDesc","LLVMSparcInfo","LLVMX86Disassembler","LLVMX86AsmParser","LLVMX86CodeGen","LLVMSelectionDAG","LLVMAsmPrinter","LLVMX86Desc","LLVMX86Info","LLVMX86AsmPrinter","LLVMX86Utils","LLVMMCDisassembler","LLVMMCParser","LLVMInterpreter","LLVMMCJIT","LLVMJIT","LLVMCodeGen","LLVMObjCARCOpts","LLVMScalarOpts","LLVMInstCombine","LLVMTransformUtils","LLVMipa","LLVMAnalysis","LLVMRuntimeDyld","LLVMExecutionEngine","LLVMTarget","LLVMMC","LLVMObject","LLVMCore","LLVMSupport","z","pthread","tinfo","stdc++","dl" ],
```
Comment (schancel) [inline dub.json]: Remove this file.   We need to add it to .gitignore.

File: lib2dub.d [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import std.stdio;
```
Comment (schancel) [inline lib2dub.d]: This should not be in the repo, at least not yet.  Submit a separate PR with a tool that properly generates the dub file completely.  You'll need to get the `--libdir` from LLVM as well as that can change also.

**Comment by schancel** [comment]
@UplinkCoder This PR is more stale than my dead grandma's underpants.  Please rebase and fix up your commit messages or close the PR.

**Comment by schancel** [comment]
@UplinkCoder I think you did something weird with your rebasing here due :)

**Comment by UplinkCoder** [comment]
@schancel your grandma just got brand-new pants

File: src/d/llvm/backend.d [hunk 3]
```diff
@@ -37,8 +39,11 @@ final class LLVMBackend {
...
-		
-		pass = new CodeGenPass(context, name);
+
+		version (D_LP64) bitWidth = 64;
+		else bitWidth = 32;
+
+		pass = new CodeGenPass(context, name, bitWidth);
 		
 		char* errorPtr;
 		auto creationError = LLVMCreateJITCompilerForModule(&executionEngine, pass.dmodule, 0, &errorPtr);
```
Comment (schancel) [inline src/d/llvm/backend.d]: See if you can't base this on my MCJIT branch.  If you do that, you can remove the executionEngine from the backend completely it looks like as you have a different way of getting the targetMachine.

**Comment by deadalnix** [comment]
This needs to be rebased on top of MCJIT. Also, if we add a switch, we need to maintain it. @UplinkCoder are you willing to do it ? If not, there is no point in adding it. people will try to use it, it won't work and we will look bad.

**Comment by UplinkCoder** [comment]
sure I will maintain it.
I'm planning to stick with sdc-devlopment.

**Comment by UplinkCoder** [comment]
rebased and tested - no regressions on 64bit

**Comment by UplinkCoder** [comment]
WHAT ???
why do I have to rebase again ?

**Comment by deadalnix** [comment]
Because I won't wait. Glancing at the code, there are various places where it is not formatted properly, because you could get the size of the pointer from target data, or at least assert that they match with what you have and because there is no test plan whatsoever.

**Comment by UplinkCoder** [comment]
Since we are on LLVM 3.5 I will change it to use the target-default for pointers

**Comment by UplinkCoder** [comment]
@deadalnix mergable now
As expected :
`Summary: 176 tests, 171 passes, 5 failures, 97.16% pass rate, 0 regressions, 0 improvements.`

File: src/d/llvm/string.d [hunk 4]
```diff
@@ -28,11 +29,11 @@ final class StringGen {
 			LLVMSetLinkage(globalVar, LLVMLinkage.Private);
 			LLVMSetGlobalConstant(globalVar, true);
 			
-			auto length = LLVMConstInt(LLVMInt64TypeInContext(pass.llvmCtx), str.length, false);
+			auto length = LLVMConstInt(LLVMIntPtrTypeInContext(pass.llvmCtx, pass.targetData), str.length, false);
 			
 			/*
 			// skip 0 termination.
```
Comment (schancel) [inline src/d/llvm/string.d]: Are these commented out lines still relevant?

File: src/d/llvm/evaluator.d [hunk 5]
```diff
@@ -211,7 +212,7 @@ final class LLVMEvaluator : Evaluator {
 		// FIXME This only works for 64 bit platforms because the retval
 		// of the "__ctfe" is specifically a i64.  This is due to MCJIT
 		// not supporting pointer return values directly at this time. 
-		auto sAsInt = LLVMGenericValueToInt(result, false);
+		auto sAsInt = cast(size_t) LLVMGenericValueToInt(result, false);
```
Comment (schancel) [inline src/d/llvm/evaluator.d]: Is the comment above still valid?

Comment (schancel) [inline src/d/llvm/evaluator.d]: Also,  what concerns from LLVM do we have if we are targeting another platform, but want to run the JIT in the current machine?

**Comment by schancel** [comment]
@UplinkCoder Looks like a good change.   Please address my comments.

**Comment by schancel** [comment]
Where are the changes related to changing the target machine to 32 bit?  I should be able to test these changes on my 64 bit machine with the right flag.  Correct?

**Comment by deadalnix** [comment]
@schancel The triple mention x86_64 , but it should be changed to match the 32 bits version of the triple.

@UplinkCoder So, the change is to build a 32bits version of SDC, but that 32 bits version still build 64 bits executables ?

**Comment by UplinkCoder** [comment]
This change is just preparation.
X86 and X86_64 are two different targets.
just changing the triple is not enough.
you need to do `LLVMGetNextTarget(LLVMGetFirstTarget())` to get X86

If this is pulled nothing will change if the targetMachine is build with x86_64 as target.

**Comment by UplinkCoder** [comment]
I'd like to introduce the actual switch in another PR.

File: src/d/llvm/evaluator.d (hunk 5 above)
Comment (UplinkCoder) [inline src/d/llvm/evaluator.d]: cross compiling will be quite hard to do with our approach!

File: src/d/llvm/string.d (hunk 4 above)
Comment (UplinkCoder) [inline src/d/llvm/string.d]: probably not...
but it's not in the scope of this pr.

**Comment by UplinkCoder** [comment]
obsoleted by https://github.com/deadalnix/SDC/pull/149
